dask/optimization_ast.py

```diff
@@ -0,0 +1,411 @@
+        self.dsk = {
+            k: v for k, v in self.dsk.items()
+            if k not in self.delete_keys
+        }
+        self.dsk[key] = (func, ) + tuple(self.arg_keys)
+
+
+    def _traverse(self, key):
+        # Avoid try... except Keyerror to keep stack traces clean
+        if key in self.dsk_key_map:
+            name = self.dsk_key_map[key]
+        else:
+            name = self._unique_name(key)
+            self.dsk_key_map[key] = name
+            val = self.dsk[key]
+            # Stop recursion when a ASTFunction is found
+            if (isinstance(val, tuple) and val and
+                    isinstance(val[0], ASTFunction)):
+                self.args.append(name)
+                self.arg_keys.append(key)
+            else:
+                ast_code = ast.Assign([ast.Name(name, ast.Store())],
+                                      self._to_ast(val))
+                self.assigns.append(ast_code)
+                self.delete_keys.add(key)
+
+        return ast.Name(name, ast.Load())
+
+
+    def _unique_name(self, obj):
+        if isinstance(obj, tuple):
+            name = "dsk_" + obj[0].split('-')[0]
+        else:
+            try:
+                name = obj.__name__
+            except AttributeError:
+                name = 'noname'
+
+        name = re.sub(r'[^a-zA-Z0-9_]', '_', name)
+        if name not in self.name_counters:
+            self.name_counters[name] = 0
+            return name
+        self.name_counters[name] += 1
+        return name + '_' + str(self.name_counters[name])
+
+
+    def _to_ast(self, v):
+        # print(f"to_ast({v})")
+        vtype = type(v)
+
+        if vtype in (int, float):
+            return ast.Num(v)
+        if vtype is bool or v is None:
+            return ast.NameConstant(v)
+        if vtype is bytes:
+            return ast.Bytes(v)
+        if vtype is str:
+            # Is it a reference to another dask graph node?
+            if v in self.dsk:
+                return self._traverse(v)
+            return ast.Str(v)
+        if vtype is set:
+            return ast.Set([self._to_ast(x) for x in v])
+        if vtype is list:
+            return ast.List([self._to_ast(x) for x in v], ast.Load())
+        if vtype is dict:
+            return ast.Dict([self._to_ast(k) for k in v.keys()],
+                            [self._to_ast(k) for k in v.values()])
+        if vtype is tuple:
+            # Is it a reference to another dask graph node?
+            try:
+                is_key = v in self.dsk
+            except TypeError:
+                # Unhashable
+                pass
+            else:
+                if is_key:
+                    return self._traverse(v)
+
+            if v and callable(v[0]):
+                return self._dsk_function_to_ast(v[0], v[1:], {})
+
+            # Generic tuple
+            return ast.Tuple([self._to_ast(x) for x in v], ast.Load())
+
+        # Additional objects explicitly handled for convenience
```

Comment:
Sorry I am a bit new into these dask internals -- could you explain why types like dict, list must be explicitly handled? What if there are subclasses, e.g. an OrderedDict? Is a `.dsk` graph guaranteed not to have such?

Reply:
they must be handled because they need to be converted to AST.
Also, any of the builtin containers may contain a dask key, which must be handled.
Subclasses are treated as generic objects (note that I'm not using isinstance), which causes them to be passed as arguments to the function.

Reply:
I see -- the consideration is not on the dask side. It is because Python has special AST elements for basic types, thus passing them as objects would have been malformed AST.